### PR TITLE
Vis post

### PR DESCRIPTION
**Pull request by mmeendez8**

### CODE REVIEW

File: _posts/2024-04-15-video-segmentation.md [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+layout: post
+title: "Exploring Transformer-Based Methods in Video Instance Segmentation"
+subtitle: "A comparison of the latest research and innovations in Video Instance Segmentation"
+description: "Get a detailed overview of the transformative impact of transformer models on video instance segmentation. Learn about the evolution of VIS methods, the role of datasets like YouTube-VIS, and the latest research trends that are defining the future of video processing technologies. This includes detailed comparisons among leading models like VisTR, IFC, and TeViT."
+image: "/assets/images/fullsize/posts/2024-04-15-video-segmentation/thumbnail.jpg"
+selected: y
+mathjax: n
+---
+
+I had recently visited the most recent literature on Video Segmentation and I found surpised about how lost I was. If you are related with it you probably are familiar with terms such as SVOS, UVOS, VIS, Zero-Shot, One-Shot, etc. If you are not then you will probably find yourself as lost as I was a few weeks ago.
```
Comment (irabanillo91) [inline _posts/2024-04-15-video-segmentation.md]:
```suggestion
I had recently visited the most recent literature on Video Segmentation and I was startled by how lost I found myself. If you are versed in this field, you are probably familiar with terms such as SVOS, UVOS, VIS, Zero-Shot, One-Shot, etc. If you are not, then you will probably find yourself as lost as I was a few weeks ago.
```

File: _posts/2024-04-15-video-segmentation.md [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+
+My main intention was to focus solely on the specific topic I was interested in. This is Video instance Segmentation (VIS) which extends the image instance segmentation task from the image domain to the video domain. The goal is to segment object instances from a predefined category set in videos and associate the instance identities across frames. It can also be seen as a combination of instance segmentation and object tracking in videos.
```
Comment (irabanillo91) [inline _posts/2024-04-15-video-segmentation.md]:
```suggestion
My main intention was to focus solely on the specific topic I was interested in. This is Video instance Segmentation (VIS), which extends the image instance segmentation task from the image domain to the video domain. The goal is to segment object instances from a predefined category set in videos, then associate the instance identities across frames. It can also be seen as a combination of instance segmentation and object tracking in videos.
```

File: _posts/2024-04-15-video-segmentation.md [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+
+As usually I will be trying to update this post with the most recent papers and code implementations I find interesting.
+
+## Introduction to Video Segmentation
+
+Video segmentation condenses a lot of different tasks which can have multiple names. I personaly like the taxonomy used in the [Youtube-VOS Dataset](https://youtube-vos.org/dataset/) which is one of the main benchmarks in this field so I will stick with it through this post. The different tasks are:
```
Comment (irabanillo91) [inline _posts/2024-04-15-video-segmentation.md]:
```suggestion
Video segmentation condenses a lot of different tasks which can have multiple names. I personaly like the taxonomy used in the [Youtube-VOS Dataset](https://youtube-vos.org/dataset/), which is one of the main benchmarks in this field ,so I will stick with it through this post. The different tasks are:
```

File: assets/svg/posts/2024-04-15-video-segmentation/video-segmentation-types-diagram.svg
Comment (irabanillo91) [inline assets/svg/posts/2024-04-15-video-segmentation/video-segmentation-types-diagram.svg:1]: **Left column**: the front frame in the input list is not the same as the front frame in the output list. I think the output is correct, but you need to update the front frame in the input. Currently it matches the segmented first frame injected as an input. I think it should be the next frame
**Middle column**: same thing.

File: _posts/2024-04-15-video-segmentation.md [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+
+- **Video Object Segmentation (VOS)**: targets at segmenting a particular object instance throughout the entire video sequence given only the object mask of the first frame.
+- **Video Instance Segmentation (VIS)**: extends image instance segmentation to videos, aiming to segment and track object instances across frames.
+- **Referring Video Object Segmentation (RVOS)**: is a task that requires to segment a particular object instance in a video given a natural language expression that refers to the object instance.
+
+<div class="post-center-image">
```
Comment (irabanillo91) [inline _posts/2024-04-15-video-segmentation.md:25]: I added comments on the svg file below

File: _posts/2024-04-15-video-segmentation.md [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+{:refdef: class="image-caption"}
+*Diagram with the different video segmentation methods*
+{: refdef}
+
+
+## Key concepts
+
+### Input sequence length
+
+The input sequence length is crucial in video segmentation. Longer sequences provide more context for accurately segmenting objects, even during occlusion and appearance changes. However, they require more computational power and increase training and inference times. Think that transformers have a quadratic complexity with respect to the sequence length so most of the models we are going to discuss are trained with very short clips (mostly between 2 and 8 frames).
```
Comment (irabanillo91) [inline _posts/2024-04-15-video-segmentation.md]:
```suggestion
The input sequence length is crucial in video segmentation. Longer sequences provide more context for accurately segmenting objects, even through occlusions or appearance changes. However, they require more computational power and they cause an increase in training and inference times. Bear in mind that transformers have a quadratic complexity with respect to the sequence length. That's the reason why most of the models we are going to discuss here are trained with very short clips (mostly between 2 and 8 frames).
```

File: _posts/2024-04-15-video-segmentation.md [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+
+### Stride
+
+The input sequence length defines the number of frames processed in parallel. The stride is the one that determines the temporal distance between frames in the input sequence. A stride of 1 means that the input sequence is a continuous sequence of frames, while a stride of 2 means that every other frame is skipped. By increasing the stride the system can work faster because it's looking at fewer frames.
```
Comment (irabanillo91) [inline _posts/2024-04-15-video-segmentation.md]: I would rephrase for clarity

```suggestion
The input sequence length defines the number of frames processed in parallel. The stride is the one that determines the temporal distance between adjacent frames in the input sequence. A stride of 1 means that the input sequence is a continuous sequence of frames, while a stride of 2 means that every other frame is skipped. By increasing the stride, the system can work faster because even though it will be looking at the same amount of frames at once, it will need to process fewer frames in total.
```

File: _posts/2024-04-15-video-segmentation.md [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+
+### Offline vs Online
+
+Many video segmentation approaches are categorized as offline, analyzing the entire video in one go, ideal for short videos and dependent on the model's capacity for the longest clip. On the other hand online methods divide videos into overlapping chunks, and results from these segments are merged using a rule-based post-tracking method. This approach ensures continuous tracking across the video by analyzing and linking instances from overlapping segments.
```
Comment (irabanillo91) [inline _posts/2024-04-15-video-segmentation.md]: I don't understand what you mean by **_`and dependent on the model's capacity for the longest clip`_** the bold sentence. You mean they're limited by the maximum length the model can process?

```suggestion
Many video segmentation approaches are categorized as offline. They process the entire video in one go, which is ideal for short videos and dependent on the model's capacity for the longest clip. On the other hand online methods divide videos into overlapping chunks, and results from these segments are merged using a rule-based post-tracking method. This approach ensures continuous tracking across the video by processing and linking instances from overlapping segments.
```

File: _posts/2024-04-15-video-segmentation.md [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+
+
+## Video Instance Segmentation
+
+From now on I will exclusively focus on Video Instance Segmentation (VIS). Most of papers before 2020 were based on either:
```
Comment (irabanillo91) [inline _posts/2024-04-15-video-segmentation.md]:
```suggestion
From now on I will exclusively focus on Video Instance Segmentation (VIS). Most papers before 2020 were based on either:
```

File: _posts/2024-04-15-video-segmentation.md [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+
+- **Top-down approach**: following tracking-by-detection methods (you can check [this other post]({% post_url 2023-11-08-tracking-by-detection-overview %}) for more information on this topic)
+- **Bottom-up approach**: clustering embedding of pixels into objects
```
Comment (irabanillo91) [inline _posts/2024-04-15-video-segmentation.md]:
```suggestion
- **Bottom-up approach**: clustering pixel embeddings into objects.
```

File: _posts/2024-04-15-video-segmentation.md [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+
+These method suffered from different issues, and around 2020, transformer-based approaches began to appear. Most of research focused on how to throw a transformer into this problem that could equal the state-of-the-art.
```
Comment (irabanillo91) [inline _posts/2024-04-15-video-segmentation.md]:
```suggestion
These method suffered from different issues, and around 2020, transformer-based approaches began to appear. Most of the research focused on how to throw a transformer into this problem that could hold up to the state-of-the-art. 
```

File: _posts/2024-04-15-video-segmentation.md [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+From now on I will exclusively focus on Video Instance Segmentation (VIS). Most of papers before 2020 were based on either:
+
+- **Top-down approach**: following tracking-by-detection methods (you can check [this other post]({% post_url 2023-11-08-tracking-by-detection-overview %}) for more information on this topic)
+- **Bottom-up approach**: clustering embedding of pixels into objects
+
+These method suffered from different issues, and around 2020, transformer-based approaches began to appear. Most of research focused on how to throw a transformer into this problem that could equal the state-of-the-art. 
+
+### Datasets
+
+The most common dataset used for VIS is called [YouTube-VIS](https://youtube-vos.org/dataset/vis/). There a total of three diferent versions:
```
Comment (irabanillo91) [inline _posts/2024-04-15-video-segmentation.md]:
```suggestion
The most common dataset used for VIS is called [YouTube-VIS](https://youtube-vos.org/dataset/vis/). It comprises three different versions:
```

File: _posts/2024-04-15-video-segmentation.md [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+
+- **YouTube-VIS-2019**: 2,883 high-resolution YouTube videos with 40 object categories. Longest video is 1,000 frames. Longest video only contains 36 frames so it is easy to execute on offline mode.
```
Comment (irabanillo91) [inline _posts/2024-04-15-video-segmentation.md:64]:
```suggestion
- **YouTube-VIS-2019**: 2,883 high-resolution YouTube videos with 40 object categories. Longest video is 1,000 frames. Shortest? video only contains 36 frames so it is easy to execute on offline mode.
```

File: _posts/2024-04-15-video-segmentation.md [hunk 13]
```diff
@@ -0,0 +1,176 @@
...
+- **YouTube-VIS-2021**: 3,859 high-resolution YouTube video with an improved 40-category label set by merging some and adding new ones. Longer video lengths force to use a near-online approach.
+- **YouTube-VIS-2022**: not considered in this post since it more recent than the papers that are included.
```
Comment (irabanillo91) [inline _posts/2024-04-15-video-segmentation.md]:
```suggestion
- **YouTube-VIS-2022**: not considered in this post since it is more recent than the papers that are covered.
```

File: _posts/2024-04-15-video-segmentation.md [hunk 14]
```diff
@@ -0,0 +1,176 @@
...
+| TeViT  | MsgShifT   | ❌   | 68.9      | 45.9     |
+| TeViT  | MsgShifT   | ✅   | 68.9      | **46.6** |
+
+</div>
+
+{:refdef: class="image-caption"}
+*Table 1. Comparisons on YouTube-VIS-2019 dataset from TeViT paper [5]. MST indicates multi-scale training strategy. FPS measured with a single TESLA V100. Note all methods used offline evaluation for reporting metrics.*
+{: refdef}
+
+Note the differences in precision when comparing with the reported results in Youtube-VIS-2021 dataset. This is due to the increase in video sizes, which forces the mode to work in online model, processing chunks of the video that then need to be merged.
```
Comment (irabanillo91) [inline _posts/2024-04-15-video-segmentation.md]:
```suggestion
Note the differences in precision when comparing with the reported results in Youtube-VIS-2021 dataset. This is due to the increase in video sizes, which forces the model to work in online mode, processing chunks of the video that then need to be merged.
```

File: _posts/2024-04-15-video-segmentation.md [hunk 15]
```diff
@@ -0,0 +1,176 @@
...
+
+</div>
+
+{:refdef: class="image-caption"}
+*Table 2. Comparisons on YouTube-VIS-2021 dataset from TeViT paper [5].*
+{: refdef}
+
+### VisTR (2021)
+
+VisTR, short for VIS Transformer, emerged as one of the initial transformer-based VIS methods to achieve notable accuracy on the YouTube-VIS dataset, thanks to an effective adaptation of DETR [2] for segmentation. This framework processes a fixed sequence of video frames using a ResNet backbone to independently extract features from each image. These extracted features are then concatenated, enriched with a 3D positional encoding, and inputted into an encoder-decoder transformer architecture, which outputs a sequence of object predictions in order.
```
Comment (irabanillo91) [inline _posts/2024-04-15-video-segmentation.md]:
```suggestion
VisTR, short for VIS Transformer, emerged as one of the initial transformer-based VIS methods to achieve notable accuracy on the YouTube-VIS dataset, thanks to an effective adaptation of DETR [2] for segmentation. This framework processes a fixed sequence of video frames using a ResNet backbone to independently extract features from each image. These extracted features are then concatenated, enriched with a 3D positional encoding, and injected into an encoder-decoder transformer architecture, which outputs a sequence of object predictions in order.
```

File: _posts/2024-04-15-video-segmentation.md [hunk 16]
```diff
@@ -0,0 +1,176 @@
...
+
+<div class="post-center-image">
+    {% picture pimage /assets/images/fullsize/posts/2024-04-15-video-segmentation/vistr.png --alt VisTR architecture diagram  %}
+</div>
+
+{:refdef: class="image-caption"}
+*VisTR architecture diagram*
+{: refdef}
+
+Things we need to highlight about this method:
```
Comment (irabanillo91) [inline _posts/2024-04-15-video-segmentation.md]:
```suggestion
Key ideas we need to highlight about this method:
```

File: _posts/2024-04-15-video-segmentation.md [hunk 17]
```diff
@@ -0,0 +1,176 @@
...
+3. The Hungarian algorithm is used to find a bipartite matching between ground-truth and prediction (as in DETR [2]).
+4. Solving 3 using masks is very expensive. A module is added to obtain bounding boxes from instance predictions and solve the matching efficiently.
+5. The loss consists on a combination of the bounding box, the mask and the class prediction errors.
+6. A 3D conv module fuses temporal information before masks are obtained.
+
+**Drawbacks**: The models are trained on 8 V100 GPUs of 32G RAM, with 1 video clip per GPU. Either low resolution or short clips are used to fit in memory. VisTR remains as a complete offline strategy because it takes the entire video as an input (from IFC paper).
+
+### IFC
+
+Inter-frame Communication Transformers (IFC) leverage the idea that since humans can summarize scenes briefly and consecutive frames often share similarities, it's feasible to communicate frame differences with minimal data. To reduce computational load, IFC utilizes a number of 'memory tokens' to exchange information between frames, thus lowering the complexity of space-time attention.
```
Comment (irabanillo91) [inline _posts/2024-04-15-video-segmentation.md]:
```suggestion
Inter-frame Communication Transformers (IFC) leverages the idea that, since humans can summarize scenes briefly and consecutive frames often share similarities, it's feasible to communicate frame differences with minimal data. To reduce computational load, IFC utilizes a number of 'memory tokens' to exchange information between frames, thus lowering the complexity of space-time attention.
```

File: _posts/2024-04-15-video-segmentation.md [hunk 18]
```diff
@@ -0,0 +1,176 @@
...
+
+<div class="post-center-image">
+    {% picture pimage /assets/images/fullsize/posts/2024-04-15-video-segmentation/tevit.png --alt TeViT architecture diagram  %}
+</div>
+
+{:refdef: class="image-caption"}
+*TeViT architecture diagram*
+{: refdef}
+
+At its core, TeViT employs a pyramid vision transformer [6] structure and innovates by replacing IFC's memory tokens with temporal messenger tokens. These tokens are periodically shifted along the temporal axis within each block to merge information from distinct frames. This shift operation is straightforward yet remarkably effective, adding no extra parameters to the system.
```
Comment (irabanillo91) [inline _posts/2024-04-15-video-segmentation.md]:
```suggestion
At its core, TeViT employs a pyramid vision transformer [6] structure and innovates by replacing IFC's memory tokens with temporal messenger tokens. These tokens are periodically shifted along the temporal axis within each block to merge information from distinct frames. This shift operation is straightforward, yet remarkably effective, adding no extra parameters to the system.
```

File: _posts/2024-04-15-video-segmentation.md [hunk 19]
```diff
@@ -0,0 +1,176 @@
...
+
+The head implementation emphasizes modeling temporal relations at the instance level, drawing on the principles of QueryInst [7]. As illustrated in the diagram, the same instances queries are initially applied across every frame. These queries are processed through a parameter-shared multi-head self-attention (MHSA) mechanism and a dynamic convolutional layer [8], which integrates the data with instance region features from the backbone. Finally, task-specific heads (such as classification, box, and mask heads) generate predictions for a sequence of video instances.
```
Comment (irabanillo91) [inline _posts/2024-04-15-video-segmentation.md]:
```suggestion
The head implementation emphasizes modeling temporal relations at the instance level, drawing on the principles of QueryInst [7]. As illustrated in the diagram, the same instance queries are initially applied across every frame. These queries are processed through a parameter-shared multi-head self-attention (MHSA) mechanism and a dynamic convolutional layer [8], which integrates the data with instance region features from the backbone. Finally, task-specific heads (such as classification, box, and mask heads) generate predictions for a sequence of video instances.
```

File: _posts/2024-04-15-video-segmentation.md [hunk 20]
```diff
@@ -0,0 +1,176 @@
...
+
+The loss computation incorporates the Hungarian algorithm alongside a combination of box, mask, and class prediction errors (details provided in the paper).
+
+
```
Comment (irabanillo91) [inline _posts/2024-04-15-video-segmentation.md:171]: Maybe a final Conclusions section?

File: _posts/2024-04-15-video-segmentation.md (hunk 7 above)
Comment (mmeendez8) [reply _posts/2024-04-15-video-segmentation.md]: Yup. But do not really like it either, I will rephrase this

File: _posts/2024-04-15-video-segmentation.md (hunk 12 above)
Comment (mmeendez8) [reply _posts/2024-04-15-video-segmentation.md:64]: longest!
